dbms/src/Functions/FunctionsGeo.cpp

```diff
@@ -246,6 +251,181 @@ class FunctionPointInPolygon : public IFunction
+
+            const size_t encoded_size = GeoUtils::geohashEncode(latitude, longitude, precision, pos);
+
+            pos += encoded_size;
+            *pos = '\0';
+            out_offsets[i] = ++pos - begin;
+        }
+        out_vec.resize(pos - begin);
+
+        if (!out_offsets.empty() && out_offsets.back() != out_vec.size())
+            throw Exception("Column size mismatch (internal logical error)", ErrorCodes::LOGICAL_ERROR);
+
+        result = std::move(col_str);
+
+        return true;
+
+    }
+
+    template <typename LatType, typename LonType>
+    bool tryExecute(const IColumn * lat_column, const IColumn * lon_column, const IColumn * precision_col, ColumnPtr & result)
+    {
+        const ColumnVector<LatType> * latitude = checkAndGetColumnEvenIfConst<ColumnVector<LatType>>(lat_column);
+        const ColumnVector<LonType> * longitude = checkAndGetColumnEvenIfConst<ColumnVector<LonType>>(lon_column);
+        if (!latitude || !longitude)
+            return false;
+
+        return tryExecute2<UInt8>(latitude, longitude, precision_col, result) ||
+               tryExecute2<UInt16>(latitude, longitude, precision_col, result) ||
+               tryExecute2<UInt32>(latitude, longitude, precision_col, result) ||
+               tryExecute2<UInt64>(latitude, longitude, precision_col, result) ||
+               tryExecute2<Int8>(latitude, longitude, precision_col, result) ||
+               tryExecute2<Int16>(latitude, longitude, precision_col, result) ||
+               tryExecute2<Int32>(latitude, longitude, precision_col, result) ||
+               tryExecute2<Int64>(latitude, longitude, precision_col, result);
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t /*input_rows_count*/) override
+    {
+        const IColumn * latitude = block.getByPosition(arguments[0]).column.get();
+        const IColumn * longitude = block.getByPosition(arguments[1]).column.get();
+        const IColumn * precision = block.getByPosition(arguments[2]).column.get();
+        ColumnPtr & res_column = block.getByPosition(result).column;
+
+        if (tryExecute<Float32, Float32>(latitude, longitude, precision, res_column) ||
+            tryExecute<Float64, Float32>(latitude, longitude, precision, res_column) ||
+            tryExecute<Float32, Float64>(latitude, longitude, precision, res_column) ||
+            tryExecute<Float64, Float64>(latitude, longitude, precision, res_column))
+            return;
+
+        throw Exception("Unsupported argument types: "
+                        + block.getByPosition(arguments[0]).column->getName() + ", "
+                        + block.getByPosition(arguments[1]).column->getName() + ", "
+                        + block.getByPosition(arguments[2]).column->getName()
+                        + " of argument of function " + getName(),
+                        ErrorCodes::ILLEGAL_COLUMN);
+    }
+};
+
+// geohashDecode(string) => (lat float64, lon float64)
```

Comment:
Could please you write documentation for these two functions and put it to https://github.com/yandex/ClickHouse/blob/master/docs/en/query_language/functions/geo.md ?

Reply:
Sure, will do

Reply:
Fixed

Reply:
Fixed